File: core-algorithms/3_sort/3_merge_sort/merge_sort.ts
const merge = (left: number[], right: number[]): number[] => {
  const result: number[] = [];
  let i = 0;
  let j = 0;

  while (i < left.length && j < right.length) {
    if (left[i] <= right[j]) {
      result.push(left[i]);
      i++;
    } else {
      result.push(right[j]);
      j++;
    }
  }

  // Add remaining elements
  return result.concat(left.slice(i)).concat(right.slice(j));
};

const mergeSort = (arr: number[]): number[] => {
  if (arr.length === 1) {
    return arr;
  } else if (arr.length === 2) {
    const min = Math.min(arr[0], arr[1]);
    const max = Math.max(arr[0], arr[1]);
    return [min, max];
  }

  const center = Math.floor(arr.length / 2);
  const leftArray = mergeSort(arr.slice(0, center));
  const rightArray = mergeSort(arr.slice(center));

  return merge(leftArray, rightArray);
};

const mergeSortedArray = mergeSort([2, 43, 1, 23, 54, 13, 119, 0]);

console.log(mergeSortedArray);

File: core-algorithms/3_sort/3_merge_sort/merge_sort.py
import math

def merge(listOne, listTwo):
    i = 0
    j = 0
    listMerged = []

    while(i < len(listOne) and j < len(listTwo)):
        if(listOne[i] < listTwo[j]):
            listMerged.append(listOne[i])
            i+=1
        else:
            listMerged.append(listTwo[j])
            j+=1
    
    return [*listMerged, *listOne[i:], *listTwo[j:]]
        

def mergeSort(lis):
    if(len(lis) == 1):
        return lis
    if(len(lis) == 2):
        return sorted(lis)
    
    median = math.floor(len(lis)/2)
    left = mergeSort(lis[0:median])
    right = mergeSort(lis[median:])

    return merge(left, right)


mergeSortedArray = mergeSort([2, 43, 1, 23, 54, 13, 119, 0]);

print(mergeSortedArray);